### DISCUSSION TITLE
Is it possible to develop workflows in Java instead of Groovy DSL?

### DISCUSSION THREAD

**Discussion by chhh** [nextflow-io/nextflow#2778] [Q&A]
I find it extremely difficult to develop complex scripts because I don't understand the anatomy and rules of scripts completely (due to Groovy's weird syntax and lack of tooling).

Is there any way to develop workflows in Java directly instead? With strong typing, help from IDEs etc.
I don't mind if that's a complicated process, just want to know if that's possible at all.

Thank you!

**Comment by bentsherman** [comment]
Well, Groovy is a sort of superset of Java, so most Java code is also valid in Groovy. In particular, you can declare variable types if you want to, instead of using `def`.

Aside from that, there are efforts underway to improve the Nextflow docs (especially regarding language syntax) and develop a basic linter.

**Reply by chhh** [reply]
The question is more about enabling tooling support, I guess. Is it possible to write a Nextflow script as a java class and compile it with Nextflow repo code as a dependency? Then we'd get all the tooling support from IDEs telling you what's applicable in the current context as you're developing a script.

**Reply by bentsherman** [reply]
I don't think you can write a Nextflow workflow as a Java or Groovy class, but you might be able to setup your IDE to treat *.nf files like *.groovy files and thereby show you what packages are available. IIRC you can import classes in Nextflow just like you would in Java or Groovy. You'll definitely need Groovy either way because the DSL syntax is specific to Groovy.

I don't know if anyone has tried this before but I think it's worth trying if it would be valuable for you.
